src/documentation/programming-guide.md

```diff
@@ -900,8 +901,190 @@ odds = results.odd
 tens = results[None]  # the undeclared main output
 ```
 
-<a name="io"></a>
-<a name="running"></a>
+## <a name="io"></a>Pipeline I/O
+
+When you create a pipeline, you often need to read data from some external source, such as a file in cloud storage or a database. Likewise, you may want your pipeline to output its result data to a similar external data sink. Beam provides read and write transforms for a number of common data storage types. If you want your pipeline to read from or write to a data storage format that isn't supported by the built-in transforms, you can implement your own read and write transforms.
+
+> A guide that covers how to implement your own Beam IO transforms is in progress ([BEAM-1025](https://issues.apache.org/jira/browse/BEAM-1025)).
+
+### Reading Input Data
+
+Read transforms read data from an external source and return a `PCollection` representation of the data for use by your pipeline. You can use a read transform at any point while constructing your pipeline to create a new `PCollection`, though it will be most common at the start of your program. Here are examples of two common ways to read data.
+
+#### Reading from a `Source`:
+
+```java
+// This example uses XmlSource.
+XmlSource<String> source =   XmlSource.<String>from(file.toPath().toString())
+     .withRootElement("root")
+     .withRecordElement("record")
+     .withRecordClass(Record.class);
+PCollection<String> output = pipeline.apply(Read.from(source));
+```
+
+```python
+pipeline | beam.io.Read(beam.io.TextFileSource("protocol://path/to/some/inputData.txt"))
```

Comment:
The example here is specifically about how to read from a `Source` - TextIO does not provide a source, so it wouldn't actually help users with what we are trying to explain in this section (once we have migrated all the IOs to use Read Transform style, we'd be able to get rid of this section of the docs altogether)

I created https://issues.apache.org/jira/browse/BEAM-1242 to track the work converting the older style transforms since I'm not sure we have something tracking that. 

Note that TextIO.Read is not actually a great example itself because it uses a weird intermediate API style TextIO.Read is not standard now. TextIO.read() would be the newer style.